File: frontend/js/leave.js
/**
 * Leave Management - FIXED VERSION
 */

// ============================================
// INITIALIZE
// ============================================
async function initLeavePage() {
    console.log('[Leave] Initializing...');
    
    await loadLeaveBalance();
    await loadLeaveRequests();
    setupLeaveForm();
}

// ============================================
// LOAD BALANCE
// ============================================
async function loadLeaveBalance() {
    console.log('[Leave] Loading balance...');
    
    try {
        const result = await api.get('/leave/balance');
        
        console.log('[Leave] Balance result:', result);
        
        if (result.success && result.data) {
            const data = result.data;
            
            const quotaEl = document.getElementById('balanceQuota');
            const usedEl = document.getElementById('balanceUsed');
            const remainingEl = document.getElementById('balanceRemaining');
            
            if (quotaEl) quotaEl.textContent = (data.annual_quota || 12) + ' hari';
            if (usedEl) usedEl.textContent = (data.annual_used || 0) + ' hari';
            if (remainingEl) remainingEl.textContent = (data.annual_remaining || 12) + ' hari';
        }
        
    } catch (error) {
        console.error('[Leave] Balance error:', error);
    }
}

// ============================================
// LOAD REQUESTS
// ============================================
async function loadLeaveRequests() {
    console.log('[Leave] Loading requests...');
    
    const container = document.getElementById('leaveList');
    if (!container) return;
    
    container.innerHTML = '<p style="text-align: center; color: var(--gray-400);"><i class="fas fa-spinner fa-spin"></i> Memuat...</p>';
    
    try {
        const result = await api.get('/leave/my-requests');
        
        console.log('[Leave] Requests result:', result);
        
        if (result.success && result.data && result.data.length > 0) {
            container.innerHTML = result.data.map(req => {
                const statusClass = {
                    pending: 'warning',
                    approved: 'success',
                    rejected: 'danger',
                    cancelled: 'secondary'
                }[req.status] || 'secondary';
                
                const statusLabel = {
                    pending: 'Menunggu',
                    approved: 'Disetujui',
                    rejected: 'Ditolak',
                    cancelled: 'Dibatalkan'
                }[req.status] || req.status;
                
                return `
                    <div class="leave-item" style="display: flex; justify-content: space-between; align-items: center; padding: 15px; border-bottom: 1px solid var(--gray-100);">
                        <div class="leave-info">
                            <strong style="display: block; margin-bottom: 5px;">${getLeaveTypeName(req.leave_type)}</strong>
                            <p style="font-size: 0.85rem; color: var(--gray-500);">${req.reason}</p>
                        </div>
                        <div class="leave-dates" style="text-align: right; margin: 0 15px;">
                            <span style="display: block; font-weight: 500;">${formatDate(req.start_date)} - ${formatDate(req.end_date)}</span>
                            <span style="font-size: 0.85rem; color: var(--gray-500);">${req.total_days} hari</span>
                        </div>
                        <span class="leave-status ${statusClass}" style="padding: 6px 12px; border-radius: 20px; font-size: 0.8rem; font-weight: 600; background: var(--${statusClass}-light); color: var(--${statusClass});">
                            ${statusLabel}
                        </span>
                    </div>
                `;
            }).join('');
        } else {
            container.innerHTML = '<p style="text-align: center; color: var(--gray-400); padding: 40px;">Belum ada pengajuan cuti</p>';
        }
        
    } catch (error) {
        console.error('[Leave] Requests error:', error);
        container.innerHTML = '<p style="text-align: center; color: var(--danger);">Gagal memuat data</p>';
    }
}

function getLeaveTypeName(type) {
    const types = {
        annual: 'Cuti Tahunan',
        sick: 'Sakit',
        marriage: 'Cuti Menikah',
        maternity: 'Cuti Melahirkan',
        paternity: 'Cuti Ayah',
        bereavement: 'Duka Cita',
        unpaid: 'Tanpa Gaji',
        other: 'Lainnya'
    };
    return types[type] || type;
}

// ============================================
// FORM SETUP
// ============================================
function setupLeaveForm() {
    const today = new Date().toISOString().split('T')[0];
    
    const startDate = document.getElementById('startDate');
    const endDate = document.getElementById('endDate');
    
    if (startDate) startDate.min = today;
    if (endDate) endDate.min = today;
    
    if (startDate) {
        startDate.addEventListener('change', () => {
            if (endDate) endDate.min = startDate.value;
        });
    }
    
    const form = document.getElementById('leaveForm');
    if (form) {
        form.addEventListener('submit', handleLeaveSubmit);
    }
}

async function handleLeaveSubmit(e) {
    e.preventDefault();
    
    console.log('[Leave] Submitting request...');
    
    const btn = e.target.querySelector('button[type="submit"]');
    if (btn) {
        btn.disabled = true;
        btn.innerHTML = '<i class="fas fa-spinner fa-spin"></i> Memproses...';
    }
    
    try {
        const data = {
            leave_type: document.getElementById('leaveType').value,
            start_date: document.getElementById('startDate').value,
            end_date: document.getElementById('endDate').value,
            reason: document.getElementById('reason').value
        };
        
        console.log('[Leave] Submit data:', data);
        
        const result = await api.post('/leave/request', data);
        
        console.log('[Leave] Submit result:', result);
        
        if (result.success) {
            showToast('Pengajuan cuti berhasil', 'success');
            closeLeaveModal();
            document.getElementById('leaveForm').reset();
            await loadLeaveRequests();
            await loadLeaveBalance();
        } else {
            showToast(result.message || 'Gagal mengajukan cuti', 'error');
        }
        
    } catch (error) {
        console.error('[Leave] Submit error:', error);
        showToast('Terjadi kesalahan', 'error');
    } finally {
        if (btn) {
            btn.disabled = false;
            btn.innerHTML = '<i class="fas fa-paper-plane"></i> Ajukan';
        }
    }
}

// ============================================
// MODAL
// ============================================
function openLeaveModal() {
    const modal = document.getElementById('leaveModal');
    if (modal) {
        modal.classList.add('show');
        document.body.style.overflow = 'hidden';
    }
}

function closeLeaveModal() {
    const modal = document.getElementById('leaveModal');
    if (modal) {
        modal.classList.remove('show');
        document.body.style.overflow = '';
    }
}

// Close on background click
document.addEventListener('click', (e) => {
    if (e.target.classList.contains('modal')) {
        closeLeaveModal();
    }
});

// ============================================
// EXPORT GLOBAL
// ============================================
window.initLeavePage = initLeavePage;
window.openLeaveModal = openLeaveModal;
window.closeLeaveModal = closeLeaveModal;

console.log('[Leave] Loaded successfully');

File: backend/routes/leave.py
from flask import request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from datetime import datetime, timedelta
from models import db, Employee, LeaveRequest, LeaveBalance
from routes import leave_bp
from utils.helpers import get_wib_now, get_wib_today

LEAVE_TYPES = {
    'annual': {'name': 'Cuti Tahunan', 'max_days': 12},
    'sick': {'name': 'Sakit', 'max_days': 14},
    'marriage': {'name': 'Cuti Menikah', 'max_days': 3},
    'maternity': {'name': 'Cuti Melahirkan', 'max_days': 90},
    'paternity': {'name': 'Cuti Ayah', 'max_days': 2},
    'bereavement': {'name': 'Duka Cita', 'max_days': 2},
    'unpaid': {'name': 'Tanpa Gaji', 'max_days': 30},
}

@leave_bp.route('/types', methods=['GET'])
@jwt_required()
def get_types():
    return jsonify({'success': True, 'data': LEAVE_TYPES}), 200

@leave_bp.route('/balance', methods=['GET'])
@jwt_required()
def get_balance():
    try:
        employee_id = get_jwt_identity()
        year = request.args.get('year', datetime.now().year, type=int)
        
        balance = LeaveBalance.query.filter_by(employee_id=employee_id, year=year).first()
        if not balance:
            balance = LeaveBalance(employee_id=employee_id, year=year)
            db.session.add(balance)
            db.session.commit()
        
        return jsonify({
            'success': True,
            'data': {
                'year': balance.year,
                'annual_quota': balance.annual_quota,
                'annual_used': balance.annual_used,
                'annual_remaining': balance.annual_remaining
            }
        }), 200
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)}), 500

@leave_bp.route('/request', methods=['POST'])
@jwt_required()
def create_request():
    try:
        employee_id = get_jwt_identity()
        data = request.get_json()
        
        start_date = datetime.strptime(data['start_date'], '%Y-%m-%d').date()
        end_date = datetime.strptime(data['end_date'], '%Y-%m-%d').date()
        
        total_days = 0
        current = start_date
        while current <= end_date:
            if current.weekday() < 5:
                total_days += 1
            current += timedelta(days=1)
        
        leave_request = LeaveRequest(
            employee_id=employee_id,
            leave_type=data['leave_type'],
            start_date=start_date,
            end_date=end_date,
            total_days=total_days,
            reason=data['reason'],
            status='pending'
        )
        db.session.add(leave_request)
        db.session.commit()
        
        return jsonify({'success': True, 'message': 'Pengajuan berhasil', 'data': leave_request.to_dict()}), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'success': False, 'message': str(e)}), 500

@leave_bp.route('/my-requests', methods=['GET'])
@jwt_required()
def get_my_requests():
    try:
        employee_id = get_jwt_identity()
        requests = LeaveRequest.query.filter_by(employee_id=employee_id).order_by(LeaveRequest.created_at.desc()).all()
        return jsonify({'success': True, 'data': [r.to_dict() for r in requests]}), 200
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)}), 500

@leave_bp.route('/cancel/<int:leave_id>', methods=['POST'])
@jwt_required()
def cancel_request(leave_id):
    try:
        employee_id = get_jwt_identity()
        leave = LeaveRequest.query.filter_by(id=leave_id, employee_id=employee_id).first()
        
        if not leave:
            return jsonify({'success': False, 'message': 'Tidak ditemukan'}), 404
        
        if leave.status != 'pending':
            return jsonify({'success': False, 'message': 'Hanya pending yang bisa dibatalkan'}), 400
        
        leave.status = 'cancelled'
        db.session.commit()
        
        return jsonify({'success': True, 'message': 'Dibatalkan'}), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({'success': False, 'message': str(e)}), 500
